packages/node_modules/cerebral/src/utils.js

```diff
@@ -277,33 +277,24 @@ export function createDummyController(state = {}, signals = {}) {
 }
 
 export function copyAndFreezeObject(rootObject) {
-  function makeCopy(value) {
-    if (Array.isArray(value)) {
-      return value.slice()
-    } else if (isObject(value)) {
-      return Object.assign({}, value)
-    }
-
-    return value
+  function deepFreeze(val) {
+    if (isObject(val)) return deepFreezeObject(val)
+    if (Array.isArray(val)) return deepFreezeArray(val)
+    return val
   }
-
-  function freezeObject(object) {
-    if (!Object.isFrozen(object) && isComplexObject(object)) {
-      for (const key in object) {
-        // Properties might not be writable, but then there
-        // is not reason to freeze its value either
-        try {
-          object[key] = freezeObject(object[key])
-        } catch (e) {}
-      }
-
-      return Object.freeze(makeCopy(object))
-    }
-
-    return object
+  function deepFreezeObject(obj) {
+    Object.getOwnPropertyNames(obj).forEach(name => {
+      deepFreeze(obj[name])
+    })
+    return Object.freeze(obj)
   }
-
-  return freezeObject(rootObject)
+  function deepFreezeArray(arr) {
+    arr.forEach(val => {
+      deepFreeze(val)
```

Comment:
I interpreted the MDN page about Object.freeze wrongly and I did not try it out myself in the Browser, sorry.
Since this is in place mutation, I think that the forEach is just right.